usbSerialForAndroid/src/main/java/com/hoho/android/usbserial/driver/GsmSerialDriver.java

```diff
@@ -0,0 +1,121 @@
+package com.hoho.android.usbserial.driver;
+
+import android.hardware.usb.UsbConstants;
+import android.hardware.usb.UsbDevice;
+import android.hardware.usb.UsbEndpoint;
+import android.hardware.usb.UsbInterface;
+import android.util.Log;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GsmSerialDriver implements UsbSerialDriver{
+
+    private static final String TAG = "GsmSerialDriver";
+    private final UsbDevice mDevice;
+
+    private final UsbSerialPort mPort;
+
+    private UsbInterface mControlInterface;
+    private UsbInterface mDataInterface;
+    private UsbEndpoint mControlEndpoint;
+
+    private int mControlIndex;
+    private static final int SET_CONTROL_LINE_STATE = 0x22;
+
+    @Override
+    public UsbDevice getDevice() {
+        return mDevice;
+    }
+
+    @Override
+    public List<UsbSerialPort> getPorts() {
+        return Collections.singletonList(mPort);
+    }
+
+    public GsmSerialDriver(UsbDevice mDevice) {
+        this.mDevice = mDevice;
+        mPort = new GsmSerialDriver.GsmSerialPort(mDevice, 0);
+    }
+
+    public class GsmSerialPort extends CommonUsbSerialPort {
+
+        public GsmSerialPort(UsbDevice device, int portNumber) {
+            super(device, portNumber);
+        }
+
+        @Override
+        protected void openInt() throws IOException {
+            mControlIndex = 0;
+            mControlInterface = mDevice.getInterface(0);
+            mDataInterface = mDevice.getInterface(0);
+            if (!mConnection.claimInterface(mControlInterface, true)) {
+                throw new IOException("Could not claim shared control/data interface");
+            }
+            Log.i(TAG, "endpointcount "+ mControlInterface.getEndpointCount());
+            for (int i = 0; i < mControlInterface.getEndpointCount(); ++i) {
+                UsbEndpoint ep = mControlInterface.getEndpoint(i);
+                Log.i(TAG, ep.toString() + "dir "+ ep.getDirection() + " type " + ep.getType());
+                if ((ep.getDirection() == UsbConstants.USB_DIR_IN) && (ep.getType() == UsbConstants.USB_ENDPOINT_XFER_INT)) {
```

Comment:
same logic as in CdcAcm driver, where controlEndpoint was not found. Better remove mControlInterface + mControlEndpoint completely.